File: frontend/src/lib/api/privacy.ts
import { apiFetch } from "@/lib/api";

export const AI_NOTICE_VERSION = "2026-09-13";
export const AI_NOTICE_STORAGE_KEY = `langy-ai-notice-${AI_NOTICE_VERSION}`;
export const MIC_NOTICE_VERSION = "2026-09-13";
export const MIC_NOTICE_STORAGE_KEY = `langy-microphone-notice-${MIC_NOTICE_VERSION}`;

export function acknowledgePrivacyNotice(
  token: string,
  notice_key: "ai_first_use" | "microphone",
  notice_version: string
) {
  return apiFetch<{ ok: boolean }>("/api/privacy/notices/acknowledge", {
    method: "POST",
    token,
    body: { notice_key, notice_version },
  });
}

export function exportPersonalData(token: string) {
  return apiFetch<Record<string, unknown>>("/api/privacy/export", { token });
}

export function deleteAccount(token: string, confirmation: string) {
  return apiFetch<{ ok: boolean }>("/api/privacy/account", {
    method: "DELETE",
    token,
    body: { confirmation },
  });
}

File: backend/app/api/routes/privacy.py
from typing import Annotated, Literal

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from sqlalchemy.orm import Session

from app.auth.deps import get_approved_user
from app.config import get_settings
from app.db import get_db
from app.domain.privacy.service import (
    acknowledge_notice,
    delete_account,
    export_user_data,
)
from app.models import User

router = APIRouter()


class NoticeAcknowledgementRequest(BaseModel):
    notice_key: Literal["ai_first_use", "microphone"]
    notice_version: str


class DeleteAccountRequest(BaseModel):
    confirmation: str


@router.post("/notices/acknowledge")
def acknowledge(
    body: NoticeAcknowledgementRequest,
    user: Annotated[User, Depends(get_approved_user)],
    db: Annotated[Session, Depends(get_db)],
) -> dict:
    acknowledge_notice(db, user, body.notice_key, body.notice_version)
    return {"ok": True}


@router.get("/export")
def export_data(
    user: Annotated[User, Depends(get_approved_user)],
    db: Annotated[Session, Depends(get_db)],
) -> dict:
    return export_user_data(db, user)


@router.delete("/account")
def remove_account(
    body: DeleteAccountRequest,
    user: Annotated[User, Depends(get_approved_user)],
    db: Annotated[Session, Depends(get_db)],
) -> dict:
    if body.confirmation != "DELETE":
        raise HTTPException(
            status_code=400, detail="Type DELETE to confirm account deletion"
        )
    try:
        delete_account(db, user, get_settings())
    except RuntimeError as exc:
        raise HTTPException(status_code=503, detail=str(exc)) from exc
    return {"ok": True}
